pyproject_save_files.py

```diff
@@ -0,0 +1,312 @@
+            "docs": [],  # to be used once there is upstream way to recognize READMEs
+            "licenses": [],  # to be used once there is upstream way to recognize LICENSEs
+        },
+        "modules": defaultdict(list),  # each importable module (directory, .py, .so)
+        "executables": {"files": []},  # regular %file entries in %{_bindir}
+        "other": {"files": []},  # regular %file entries we could not parse :(
+    }
+
+    # Note that there are no directories, only files !  # TODO find documentation
+    for path in parsed_record_content:
+        if path.suffix == ".pyc":
+            # we handle bytecode separately
+            continue
+
+        if path.parent == distinfo:
+            # TODO is this a license/documentation?
+            paths["metadata"]["files"].append(path)
+            continue
+
+        if path.parent == bindir:
+            paths["executables"]["files"].append(path)
+            continue
+
+        for sitedir in sitedirs:
+            if sitedir in path.parents:
+                if path.parent == sitedir:
+                    if path.suffix == ".so":
+                        # extension modules can have 2 suffixes
+                        name = BuildrootPath(path.stem).stem
+                        add_file_to_module(paths, name, "extension", path)
+                    elif path.suffix == ".py":
+                        name = path.stem
+                        add_file_to_module(paths, name, "script", *pycached(path))
+                    else:
+                        # TODO classify .pth files
+                        warnings.warn(f"Unrecognized file: {path}")
+                        paths["other"]["files"].append(path)
+                else:
+                    # this file is inside a dir, we classify that dir
+                    index = path.parents.index(sitedir)
+                    module_dir = path.parents[index - 1]
+                    add_file_to_module(paths, module_dir.name, "package", module_dir)
+                break
+        else:
+            warnings.warn(f"Unrecognized file: {path}")
+            paths["other"]["files"].append(path)
+
+    return paths
+
+
+def generate_file_list(paths_dict, module_globs, include_executables=False):
+    """
+    This function takes the classified paths_dict and turns it into lines
+    for the %files section. Returns list with text lines, no Path objects.
+
+    Only includes files from modules that match module_globs, metadata and
+    optional executables.
+    """
+    files = set()
+
+    if include_executables:
+        files.update(f"{p}" for p in paths_dict["executables"]["files"])
+
+    files.update(f"{p}" for p in paths_dict["metadata"]["files"])
+    for macro in "dir", "doc", "license":
+        files.update(f"%{macro} {p}" for p in paths_dict["metadata"][f"{macro}s"])
+
+    modules = paths_dict["modules"]
+
+    for name in modules:
+        for glob in module_globs:
+            if fnmatch.fnmatchcase(name, glob):
```

Comment:
A discussion about invalid identifier names got me thinking.

I think we should fail if the glob matches nothing. Now we silently skip such globs.

Reply:
Added to TODO.